instant-xml/src/lib.rs

```diff
@@ -181,8 +200,163 @@ pub struct FieldContext<'xml> {
 }
 
+pub enum EntityType {
+    Element,
+    Attribute,
+}
+
+pub enum XMLTagName<'xml> {
+    FieldName,
+    Custom(&'xml str),
+}
+
 pub trait FromXml<'xml>: Sized {
-    fn from_xml(input: &str) -> Result<Self, Error>;
+    const TAG_NAME: XMLTagName<'xml>;
+
+    fn from_xml(_input: &str) -> Result<Self> {
+        unimplemented!();
+    }
+
+    fn deserialize(deserializer: &mut Deserializer, kind: EntityType) -> Result<Self>;
+}
+
+pub trait Visitor<'xml>: Sized {
+    type Value;
+
+    fn visit_str(self, _value: &str) -> Result<Self::Value> {
+        unimplemented!();
+    }
+
+    fn visit_struct<'a>(&self, _deserializer: &'a mut Deserializer) -> Result<Self::Value> {
+        unimplemented!();
+    }
+}
+
+pub struct Deserializer<'xml> {
+    parser: XmlParser<'xml>,
+    namespaces: HashMap<&'xml str, &'xml str>,
+    tag_attributes: Vec<(String, String)>,
+}
+
+impl<'xml> Deserializer<'xml> {
+    pub fn new(input: &'xml str) -> Self {
+        Self {
+            parser: XmlParser::new(input),
+            namespaces: std::collections::HashMap::new(),
+            tag_attributes: Vec::new(),
+        }
+    }
+
+    pub fn peek_next_tag(&mut self) -> Result<Option<XmlRecord>> {
+        self.parser.peek_next_tag()
+    }
+
+    pub fn verify_namespace(&self, namespace_to_verify: &str) -> bool {
+        self.namespaces.get(namespace_to_verify).is_some()
+    }
+
+    pub fn peek_next_attribute(&self) -> Option<&(String, String)> {
+        self.tag_attributes.last()
+    }
+
+    pub fn deserialize_struct<V>(
+        &mut self,
+        visitor: V,
+        name: &str,
+        namespaces: &HashMap<&'xml str, &'xml str>,
+    ) -> Result<V::Value>
+    where
+        V: Visitor<'xml>,
+    {
+        let new_namespaces = namespaces
+            .iter()
+            .filter(|(k, v)| self.namespaces.insert(k, v).is_none())
+            .collect::<Vec<_>>();
+
+        self.process_open_tag(name, namespaces)?;
+        let ret = visitor.visit_struct(self)?;
+
+        self.check_close_tag(name)?;
+        let _ = new_namespaces
+            .iter()
+            .map(|(k, _)| self.namespaces.remove(*k));
+
+        Ok(ret)
+    }
+
+    fn deserialize_bool<V>(&mut self, visitor: V) -> Result<V::Value>
+    where
+        V: Visitor<'xml>,
+    {
+        self.parser.next();
+        if let Some(item) = self.parser.next() {
+            match item? {
+                XmlRecord::Element(v) => {
+                    let ret = visitor.visit_str(v.as_str());
+                    self.parser.next();
+                    ret
+                }
+                _ => Err(Error::UnexpectedTag),
+            }
+        } else {
+            Err(Error::MissingValue)
+        }
+    }
+
+    fn deserialize_attribute<V>(&mut self, visitor: V) -> Result<V::Value>
+    where
+        V: Visitor<'xml>,
+    {
+        visitor.visit_str(&self.pop_next_attribute_value()?)
+    }
+
+    fn pop_next_attribute_value(&mut self) -> Result<String> {
```

Comment:
Can just inline this in `deserialize_attribute()`, given that it has no other callers?